Ensure frontend is available if integrations fail to start - Part 1 of 2

bdraco opened the pull request:
Part 1 of 2.  

When integrations configured via the UI block startup or fail to start,
the webserver can remain offline which make it is impossible
to recover without manually changing files in
`.storage` since the UI is not available.

This change is the foundation that part 2 will build on
and enable a listener to start the webserver when the frontend
is finished loading.

Frontend Changes (https://github.com/home-assistant/frontend/pull/6068)

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Load the frontend and hassio earlier in the bootstrap process.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #35924
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -43,6 +45,7 @@
     # To provide account link implementations
     "cloud",
 }
+FRONTEND_INTEGRATIONS = {"frontend"}
```
Why not just add it to stage 1 ?

bdraco replied:
Why not indeed.. done!

balloob commented on homeassistant/components/http/view.py:
```diff
@@ -103,8 +103,11 @@ def request_handler_factory(view, handler):
+
+        # We do not want to serve http requests when
+        # stopping.
+        if not request.app[KEY_HASS].is_setup:
```
Why do we need a `setup` phase? We can start serving once the `frontend` integration is loaded or when `EVENT_HOMEASSISTANT_START` is fired, whichever comes first.

Starting the HTTP server early is going to be interesting. I think that we can do without a blocking spinner or a reload, since we're subscription based.

balloob replied:
Only generateLovelaceUI might be impacted, but only if the user is _opening_ the frontend during startup. If it was already open, it would just show a bunch of "entity not found" but that's fine.

bdraco replied:
I needed a way to know that `frontend` has finished loaded in order to start the webserver as everything breaks if the frontend isn't fully loaded because `IndexView` and other are missing. Is there an event fired when `frontend` is finished loading ?

bdraco replied:
```
2020-05-24 21:15:17 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=frontend>
```

Answering my own question here.

bdraco replied:
Removed the setup phase.

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -42,6 +44,10 @@
+    # Get the frontend up and running as soon
+    # as possible so problem integrations can
+    # be removed
+    "frontend",
```
We also need `config` or else we don't have APIs to manage the config entries.

```suggestion
    "frontend",
    "config",
```

balloob commented:
@bramkragten this change will need your input and testing. How much will break if we allow the frontend to connect before startup is finished.

bdraco commented:
github being github
`error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.`

bdraco commented:
/AzurePipelines Run

bdraco commented on homeassistant/bootstrap.py:
```diff
@@ -421,5 +428,11 @@ async def async_setup_multi_components(domains: Set[str]) -> None:
 
         await async_setup_multi_components(stage_2_domains)
 
+    # The websocket has already started.
```
I'm not sure this is safe to do, but I haven't come up with another way to only get the startup tasks

bdraco replied:
Didn't solve it either.

bdraco replied:
Removing it solves it and everything appears to work as expected.  I'm concerned it might mean we fire the started event too early though.

bdraco replied:
Not tracking the websocket tasks seems to have solved this.  I put the block back in with a timeout just to be sure.

bdraco commented:
Looks like the UI doesn't seem to need a reload
~~A change is likely needed on the frontend to reload once the `homeassistant_start` event is seen.~~

bdraco commented:
```
***** ERROR
Tests are leaving files behind. Please update the tests to avoid writing any files:
tests/testing_config/.storage/http

```

bdraco commented:
```
except asyncio.TimeoutError:
439 |   | _LOGGER.warning(
```
Not covered by the patch.

bdraco commented:
Diff hit up to 100%

Still need to figure out

```
***** ERROR
Tests are leaving files behind. Please update the tests to avoid writing any files:
tests/testing_config/.storage/http
```

pvizeli reviewed (changes requested):
We need first change the supervisor and add a new function that the supervisor know that HA is up and running. Means, we can not merge this until the Supervisor have integrate the new way and all instance run that supervisor in a nice way with backwards compatibility.

EDIT: I don't think that fix the root problem, it is just a workaround. Maybe we need on some point our `asyncio.wait` back to safe the flow if somethings break.

bramkragten commented:
I think loading the frontend earlier is always a win. It will make the startup feel way faster to the user.

bramkragten commented:
Things I can think off:
- Don't allow removing restored entities before complete startup
- Generated Lovelace, it can show the "you have no devices available" message, so we would have to catch that when we haven't complete startup and regenerate when startup is completed. But I also see it restoring entities slowly and finally showing just a limited amount of cards (I guess we regenerate Lovelace on connect, should move that to the started event, but only when not started yet)
- When you are on a panel on restart that is not loaded on frontend start it will not show after startup
- Probably some more stuff that will not work until after a refresh

The rest behaves pretty well because we update the UI when something changes, so you will see the sidebar, config panel, integrations build-up slowly.

balloob commented on homeassistant/core.py:
```diff
@@ -1391,6 +1400,7 @@ def as_dict(self) -> Dict:
             "version": __version__,
             "config_source": self.config_source,
             "safe_mode": self.safe_mode,
+            "start_complete": self.start_complete,
```
should this be a boolean or should we just include `hass.state` ?

bdraco replied:
`hass.state` seems like it will stand up better with time assuming that works for @bramkragten

bdraco replied:
I switched this out

balloob commented on homeassistant/components/http/__init__.py:
```diff
@@ -216,29 +218,36 @@ async def async_setup(hass, config):
+        if event.data[ATTR_COMPONENT] != "frontend":
+            return
+
+        await start_server(event)
```
Can we end up calling this twice?

balloob replied:
As in, once via this event handler and once from the start event ?

bdraco replied:
Both listeners get canceled before this is called
https://github.com/home-assistant/core/pull/36093/files#diff-829fc3a9a7a79507cc486a65c18116eeR230-R231

bdraco committed:
Ensure frontend is available if integrations fail to start

When integrations block startup or fail to start
the webserver remains offline and it is impossible
to recover without manually changing files in
`.storage` since the UI is not available.

bdraco commented:
Too many conflicts, will resquash

bdraco commented:
` Unable to serialize to JSON. Bad data found at $.result.state=RUNNING(<enum 'CoreState'>`

bdraco committed:
make json serial

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -422,4 +432,11 @@ async def async_setup_multi_components(domains: Set[str]) -> None:
     # Wrap up startup
-    await hass.async_block_till_done()
+    try:
+        async with timeout(TIMEOUT_EVENT_BOOTSTRAP):
```
This seems to do a very similar thing as #36082. I think that we should keep it out of this PR as it's already impactful enough.

bdraco replied:
I'm concerned about taking it out as the user can now interact with the api/ui and trigger something in an integration that blocks for a long time like websocket did.  Taking it out might rehash #6973

balloob commented on homeassistant/components/http/__init__.py:
```diff
@@ -216,29 +218,36 @@ async def async_setup(hass, config):
+        hass.bus.async_listen(EVENT_COMPONENT_LOADED, async_wait_frontend_load)
+    )
+    startup_listeners.append(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_server)
```
If you do listen_once, the listener will already be unsubscribed, and then unsubscribing it again in `start_server` will cause a warning. So by making this `async_listen` we avoid that.
```suggestion
        hass.bus.async_listen(EVENT_HOMEASSISTANT_START, start_server)
```

bdraco committed:
Update homeassistant/components/http/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

balloob commented on homeassistant/components/websocket_api/decorators.py:
```diff
@@ -31,7 +31,9 @@ def async_response(
     def schedule_handler(hass, connection, msg):
         """Schedule the handler."""
-        hass.async_create_task(_handle_async_response(func, hass, connection, msg))
+        # As the webserver is now started before the start
+        # event we do not want to block for websocket responders
```
This is going to be a problem for anything spawned by an API interaction by the user while Home Assistant is starting up.

bdraco replied:
I'm thinking long running logbook api calls or history api calls could block startup for a few seconds.  Is there anything else that is a problem hot spot that I'm not thinking about?

balloob replied:
Two things I can think of:
 - Triggering a script that takes X time
 - Starting a config flow for new integration that starts installation of a big requirements (like `zwave`)

balloob replied:
I don't think that it would be a big problem, eventually all tasks should stop and no websocket API should start long running tasks

bdraco committed:
Fix test mock

balloob commented:
Re Bram's earlier list. Restored entities are not written to the state machine until the start event. So that's not a problem. 

We're going to miss something and it is going to bite us. I think worst case we can always roll it back. 

We need to make sure that we do block any sync with external systems until we're fully started or else an external system will miss entities and thus throw them away. HomeKit, Alexa and Google are all controlled from within Home Assistant so we should be able to block syncing until we are started (I think we already are). However if an integration is connecting TO Home Assistant, it will miss things. How do we deal with that ? 

I guess we can tell those integrations to check for the `homeassistant_started` event. Because of race conditions, any integration waiting for Home Assistant to be started should refetch the config after the event subscription for `ha_started` is in place because it could have been fired while subscribing.

bdraco commented:
Incoming webhooks/cloudhooks could get premature updates, but I haven't found a case where I think thats an issue yet.

balloob commented:
Integrations are either taking data in and representing it as events or states or integrations provide automations. Since our automation integrations don't start until the start event, I don't think that we have to worry about webhooks.

balloob commented:
Waiting for a final check off from @pvizeli to make sure that the Supervisor health check still works as intended or that we need it tweaking before we merge this. 

For frontend, I think that we should implement:
 - panel resolver, when it cannot find a panel, should check if Home Assistant is still starting, if so, it should show a spinner "Waiting for Home Assistant to start".
 - Generated Lovelace should probably just wait until started too. It will jump around too much otherwise?

pvizeli commented:
> Waiting for a final sign off from @pvizeli to make sure that the Supervisor health check still works as intended or that we need it tweaking before we merge this.

It breaks the hole add-on flows. It's how we detect a running instance not the health check but that needs to be also tweaked. I will look after my baby break into that topic (2 months).

bdraco commented:
I'm wondering if we add `hassio` to Phase 1 it could allow us to avoid the need to make any supervisor changes.

bdraco committed:
Adjust

bdraco commented:
Moved `hassio` to phase 1.  Created an integration that blocks startup for 301 seconds.  All the panels now work on reconnect when I'm on the supervisor panel and trigger a restart

bdraco commented:
```
            # 2: Check if API response
            if await self.sys_run_in_executor(
                check_port, self.ip_address, self.api_port
            ):
                _LOGGER.info("Detect a running Home Assistant instance")
                self._error_state = False
                return
```
Looks like its a `connect()` check which will now happen much sooner

bdraco commented:
I'm not sure if its actually ever an issue, but there is still a small race window where its possible that the frontend is finished loading before `hassio`, the `frontend` loaded event fires, the webserver starts, the reconnect happens, and `hassio` hasn't finished registering its endpoints.   It probably can't actually happen now, but if we ever refactor `hassio` to take longer to load than `frontend`, it could break.

If we think this is a real risk, we could fire a `STARTING` event after all the phase1 integrations are loaded and wait for that to start the webserver.  It would likely be a few milliseconds later.

bramkragten commented:
It doesn't matter if the panel is registered after the frontend load, that will work fine, the problem is more for the add-ons and to check for a rollback I guess?

bdraco commented:
Looks like the CI failed because the tests are now taking longer than 1h which is the max time allowed by azure

bdraco commented:
> It doesn't matter if the panel is registered after the frontend load, that will work fine, the problem is more for the add-ons and to check for a rollback I guess?


- Addons seem fine, they already have to deal with Home Assistant restarting.
- Rollbacks might be an issue.  Need to better understand how they work

bramkragten commented:
I don't know, @pvizeli should be able to answer what would be the problem exactly

bdraco commented:
Rollback:

If we want to rollback if the webserver fails to start, this is the behavior we would have with this PR as-is.  This gives the user the option of fixing whatever is now blocking startup on the new version OR they should still be able to access the supervisor UI and rollback to the last image.  **I'm not sure if there will be an image for them to rollback to though as I can't figure out if we create one on update.**   

If we want to rollback if the Home Assistant start event isn't fired after 10m, we would need to modify `supervisor.homeassistant.HomeAssistant._block_till_run` to pool the api after the `connect()` check is successful.

bdraco committed:
fire EVENT_CORE_CONFIG_UPDATE as well

bdraco committed:
fix test now that EVENT_CORE_CONFIG_UPDATE is fired

bdraco committed:
remove frontend start listener

bdraco committed:
fix conflict

bdraco committed:
remove cruft

bdraco commented:
The PR has been adjusted to remove the breaking change in part 1.

This should now be mergeable without impacting supervisor as the the listener to start the webserver sooner has been removed which restores the behavior to the status quo while still supporting the frontend PR.

When we are confident the supervisor is ready we can add 
```
    startup_listeners.append(
        hass.bus.async_listen(EVENT_COMPONENT_LOADED, async_wait_frontend_load)
    )
```
back to http in a new breaking change pr.

bdraco committed:
remove the orphened function as well

bdraco committed:
remve cruft from earlier refactor

balloob commented:
Nice move on breaking it up.

MartinHjelmare commented on homeassistant/components/websocket_api/http.py:
```diff
@@ -165,7 +165,9 @@ def handle_hass_stop(event):
-        self._writer_task = self.hass.async_create_task(self._writer())
+        # As the webserver is now started before the start
+        # event we do not want to block for websocket responses
+        self._writer_task = self.hass.loop.create_task(self._writer())
```
See above.

bdraco replied:
Addressed here https://github.com/home-assistant/core/pull/36263/commits/f017f9169622f7f9c23718eafb2bf071f9e3300c

MartinHjelmare commented on homeassistant/components/websocket_api/decorators.py:
```diff
@@ -31,7 +31,9 @@ def async_response(
-        hass.async_create_task(_handle_async_response(func, hass, connection, msg))
+        # As the webserver is now started before the start
+        # event we do not want to block for websocket responders
+        hass.loop.create_task(_handle_async_response(func, hass, connection, msg))
```
We should use `asyncio.create_task` now. It's recommended from Python 3.7.

bdraco replied:
Addressed in https://github.com/home-assistant/core/pull/36263

pvizeli commented:
Just for the full story. I think that is a good approach and we should also keep it up after because it avoids that other software starts to use the API before HA is ready. Unfortunately, this works not with the current code base of the supervisor but it allows us to make a simple solution on the supervisor which works on the old and new core.
